src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourceMappingOptionsControl.xaml.cs

```diff
@@ -0,0 +1,250 @@
+            var settings = componentModelMapping.GetService<ISettings>();
+            PackageSourceMappingProvider packageSourceMappingProvider = new PackageSourceMappingProvider(settings);
+            _originalPackageSourceMappings = packageSourceMappingProvider.GetPackageSourceMappingItems();
+            IReadOnlyList<MappingUIDisplay> SourceMappingsCollectiontemp = ReadMappingsFromConfigToUI(_originalPackageSourceMappings);
+            //clear sourcemappings so that they don't repeat
+            SourceMappingsCollection.Clear();
+            foreach (MappingUIDisplay item in SourceMappingsCollectiontemp)
+            {
+                SourceMappingsCollection.Add(item);
+            }
+            //make sure all buttons show on open if there are already sourcemappings
+            (ClearButtonCommand as ButtonCommand).InvokeCanExecuteChanged();
+            (RemoveButtonCommand as ButtonCommand).InvokeCanExecuteChanged();
+        }
+        private void ExecuteShowButtonCommand(object parameter)
+        {
+            _addMappingDialog = new AddMappingDialog(this);
+            IntPtr parent = WindowHelper.GetDialogOwnerHandle();
+            WindowHelper.ShowModal(_addMappingDialog, parent);
+        }
+
+        private bool CanExecuteShowButtonCommand(object parameter)
+        {
+            return true;
+        }
+
+        private void ExecuteRemoveButtonCommand(object parameter)
+        {
+            SourceMappingsCollection.Remove((MappingUIDisplay)packageList.SelectedItem);
+            (ClearButtonCommand as ButtonCommand).InvokeCanExecuteChanged();
+        }
+
+        private bool CanExecuteRemoveButtonCommand(object parameter)
+        {
+            return SourceMappingsCollection.Count > 0;
+        }
+
+        private void ExecuteClearButtonCommand(object parameter)
+        {
+            SourceMappingsCollection.Clear();
+            (RemoveButtonCommand as ButtonCommand).InvokeCanExecuteChanged();
+        }
+
+        private bool CanExecuteClearButtonCommand(object parameter)
+        {
+            return SourceMappingsCollection.Count > 0;
+        }
+
+        internal bool ApplyChangedSettings()
+        {
+            IReadOnlyDictionary<string, IReadOnlyList<string>> Patterns = new Dictionary<string, IReadOnlyList<string>>();
+            PackageSourceMapping packageSourceMappings = new PackageSourceMapping(Patterns);
+            var componentModel = NuGetUIThreadHelper.JoinableTaskFactory.Run(ServiceLocator.GetComponentModelAsync);
+            var settings = componentModel.GetService<ISettings>();
+            PackageSourceMappingProvider packageSourceMappingProvider = new PackageSourceMappingProvider(settings);
+            IReadOnlyList<PackageSourceMappingSourceItem> packageSourceMappingsSourceItems = ReadMappingsFromUIToConfig(SourceMappingsCollection);
+            try
+            {
+                if (SourceMappingsChanged(_originalPackageSourceMappings, packageSourceMappingsSourceItems))
```

Comment:
Do we need to call this method to find out if `Source Mapping` changed?
Doesn't `packageSourceMappingProvider.SavePackageSourceMappings` method only save if there is change?

Reply:
No SavePackageSourceMappings will go through all of the mappings and call AddOrUpdate even if nothing is changed (it won't modify the config file, though). It seems like it makes more sense to check if Save even needs to be called in the first place